### DISCUSSION TITLE
Add new field data to existing CGNS file

### DISCUSSION THREAD

**Discussion by JulianWgs** [CGNS/CGNS#287] [CGNStalk]
Hello,

I want to write new field data to an existing CGNS file, so I can open the file in ParaView and visualize the new data.
(The new fields are the components of a PCA.)

I'm currently using Python, but any other tool (FOSS preferred) could be viable. I do know about the pyCGNS project, but since it is possible to open CGNS files with h5py I dont use it and I dont think it is necessary.

This codes works now. I can open the file in ParaView and view the added fields (Original code at the bottom of the post, here is the updated code):

```python
import h5py
import numpy as np

# Contains the mesh and simulation data
file = h5py.File("components.cgns", "r+")
# Contains the components of a PCA
components = np.load("components.npy")

# Iterate over first three components and write the to the CGNS file
for k in range(3):
    field_name = f"p_{k:02}"
    file.copy(
        f"/Base_Volume_Elements/internalMesh/PointData/p",
        f"/Base_Volume_Elements/internalMesh/PointData/{field_name}"
    )
    file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name][" data"][...] = components[k]
    file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name].attrs["name"] = np.bytes_(bytes(field_name, 'ascii').ljust(33, b"\x00"))
```

Thank you very much!

Best regards
Julian

Old code:

```python
import h5py
import numpy as np

# Contains the mesh and simulation data
file = h5py.File("components.cgns", "r+")
# Contains the components of a PCA
components = np.load("components.npy")

# Iterate over first three components and write the to the CGNS file
for k in range(3):
    field_name = f"p_{k:02}"
    file.copy(
        f"/Base_Volume_Elements/internalMesh/PointData/p",
        f"/Base_Volume_Elements/internalMesh/PointData/{field_name}"
    )
    file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name][" data"][...] = components[k]
    file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name]["name"] = field_name
```

**Comment by MicK7** [comment]
Hello, 
it is indeed possible to read cgns file with h5py. To write is another story, you need to ensure that 
- file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name]["name"] is a constant 33 length string null terminated
- file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name]["type"] is a constant size string null terminated
- file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name]["label"] is a constant size string null terminated

You may be interested by this file https://github.com/pyCGNS/pyCGNS/blob/main/CGNS/MAP/cgio.py

**Comment by NovA80** [comment]
Also, have a look at https://github.com/splav/pylibcgns - 1-to-1 python bindings for c-functions from libcgns.

**Comment by JulianWgs** [comment]
Thanks for the quick answers!

@MicK7 are you sure this is correct? For me it only shows up in ParaView with `file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name].attrs["name"] = np.string_(field_name.ljust(33, "\x00"))`. 

In ParaView the name display in the import prompt is "p_00Array_t", but should be "p_00", which is weird. I also found out that that ``s = "hello"; np.string_(s) == np.string_(s.ljust(33, "\x00"))`, which might have something to do with it. 

It shows up in ParaView at the import prompt, but unfortunately I cant display the field. It does not show up in the dropdown menu at the top bar. Do you know why or should I consult a ParaView forum?

PS: I've updated my code

**Reply by MicK7** [reply]
Sorry it seems the file was not well tested on python3, it should work better like this:

file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name].attrs["name"] = bytes(field_name, 'ascii').ljust(33, b"\x00")
file["Base_Volume_Elements"]["internalMesh"]["PointData"][field_name].attrs["label"] = bytes("DataArray_t", 'ascii').ljust(33, b"\x00")

Concerning ParaView not reading the data, it means the data are not correctly written.
Can you try to run cgnscheck on the written cgns hdf5 file ?
Is the numpy array passed to h5py Fortran ordered ?

**Reply by JulianWgs** [reply]
Thanks again for coming back to me so quick :)

Unfortunately the proposed code did not work. Fortunately I've now found a solution `np.bytes_(bytes(field_name, 'ascii').ljust(33, b"\x00"))`. I don't know if a shorter solution exists. Everything works fine now! I'm very happy. Thanks again!

I've also updated the code in my original question :)
